lib/charms/mysql/v0/mysql.py

```diff
@@ -1252,6 +1261,70 @@ def offline_mode_and_hidden_instance_exists(self) -> bool:
 
         return matches.group(1) != "0"
 
+    def tls_set_custom(self, ca_path: str, key_path: str, cert_path: str) -> None:
+        """Execute a list of commands to enable TLS.
+
+        Args:
+            ca_path: Path to the CA certificate
+            key_path: Path to the server key_path
+            cert_path: Path to the server certificate
+        """
+        enable_commands = (
+            f"SET PERSIST ssl_ca='{ca_path}';"
+            f"SET PERSIST ssl_key='{key_path}';"
+            f"SET PERSIST ssl_cert='{cert_path}';"
+            "SET PERSIST require_secure_transport=on;"
+            "ALTER INSTANCE RELOAD TLS;"
+        )
+
+        try:
+            self._run_mysqlcli_script(
+                enable_commands,
+                user=self.server_config_user,
+                password=self.server_config_password,
+            )
+        except MySQLClientError:
+            raise MySQLTLSSetCustomConfigError("Failed to set custom TLS configuration")
+
+    def tls_restore_default(self) -> None:
+        """Restore TLS configuration to default."""
+        restore_commands = (
+            "SET PERSIST ssl_ca='ca.pem';"
+            "SET PERSIST ssl_key='server-key.pem';"
+            "SET PERSIST ssl_cert='server-cert.pem';"
+            "SET PERSIST require_secure_transport=off;"
+            "ALTER INSTANCE RELOAD TLS;"
+        )
+
+        try:
+            self._run_mysqlcli_script(
+                restore_commands,
+                user=self.server_config_user,
+                password=self.server_config_password,
+            )
+        except MySQLClientError:
+            raise MySQLTLSRestoreDefaultConfigError
+
+    def kill_unencrypted_sessions(self) -> None:
+        """Kill non local, non system open unencrypted connections."""
+        kill_connections_command = (
+            f"shell.connect('{self.server_config_user}:{self.server_config_password}@{self.instance_address}')",
+            (
+                'processes = session.run_sql("'
+                "SELECT processlist_id FROM performance_schema.threads WHERE "
+                "connection_type = 'TCP/IP' AND type = 'FOREGROUND';"
+                '")'
+            ),
+            "process_id_list = [id[0] for id in processes.fetch_all()]",
+            'for process_id in process_id_list:\n  session.run_sql(f"KILL CONNECTION {process_id}")',
+        )
```

Comment:
would this result in client queries throwing an exception? previously, we restarted the instance (which i would expect would terminate connections when it is idle)

Reply:
Client connections will get dropped, in the same way a connection lost would do (and the same way restarting the daemon would for that matter) - so nothing changes from connected apps pov.

Reply:
I was under the impression that when we restart the daemon, it will have a grace period to handle SIGTERM (where it can gracefully close open connections). Is this incorrect? If not, then the above killing of threads is different than the restarting the daemon approach

Reply:
The gracefull stop of an open connection has no difference from killing it, since the doc states:

> [KILL CONNECTION](https://dev.mysql.com/doc/refman/8.0/en/kill.html) is the same as [KILL](https://dev.mysql.com/doc/refman/8.0/en/kill.html) with no modifier: It terminates the connection associated with the given processlist_id, after terminating any statement the connection is executing.

So it take effects only when idle